### PR TITLE
Transaction enum

### LINKED ISSUE #4157: Provide Transaction enum
Rename current `Transaction` struct to `TransactionV1` and create a Transaction enum with variants for TransactionV1 and Deploy

### PR DESCRIPTION
**Pull request by Fraser999**
This PR adds the new `Transaction` enum to `casper-types`, with the recently-introduced `Transaction` struct being renamed to `TransactionV1`.

Closes #4157.

### CODE REVIEW

**Comment by cspramit** [comment]
@EdHastingsCasperLabs to review

File: types/src/transaction/transaction_v1/errors_v1.rs [hunk 1]
```diff
@@ -10,48 +10,50 @@ use std::error::Error as StdError;
 use datasize::DataSize;
 use serde::Serialize;
 
+#[cfg(doc)]
+use super::TransactionV1;
 use crate::{crypto, TimeDiff, Timestamp};
 
-/// Returned when a `Transaction` fails validation.
+/// Returned when a [`TransactionV1`] fails validation.
```
Comment (mpapierski) [inline types/src/transaction/transaction_v1/errors_v1.rs:17]: Wouldn't

```
[`super::TransactionV1`](TransactionV1)
```
work  (or something like that)? We could avoid `#[cfg(doc)]`. It's not a big deal, just wondering

**Review by mpapierski — APPROVED** [review]
I see mostly renames and new types built on top of the previous PR, and I didn't spot any issues. LGTM

File: types/src/transaction/transaction_v1/errors_v1.rs (hunk 1 above)
Comment (Fraser999) [reply types/src/transaction/transaction_v1/errors_v1.rs:17]: Well it'd be:
```rust
[`TransactionV1`](super::TransactionV1)
```
but yes it'd be equivalent.  I've just been preferring the imports approach throughout.

**Comment by Fraser999** [comment]
bors merge

**Comment by Fraser999** [comment]
bors merge

**Pull request merged**